packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -824,9 +824,22 @@ class _TextSelectionHandleOverlayState
         break;
     }
 
+    // On some platforms we may want to calculate the start and end handles
+    // separately so they scale for the selected content.
+    //
+    // For the start handle we compute the rectangles that encompass the beginning
+    // of the selection.
+    //
+    // For the end handle we compute the rectangles that encompass the end of
+    // the selection.
+    final Rect? startHandleRect = widget.renderObject.getRectForComposingRange(TextRange(start: widget.selection.start, end: widget.selection.start + 1));
+    final Rect? endHandleRect = widget.renderObject.getRectForComposingRange(TextRange(start: widget.selection.end, end: widget.selection.end - 1));
```

Comment:
`getRectForComposingRange` returns null when (this probably needs better documentation):
- the range is invalid
- the range isCollapsed
- the range does not include a full cluster (see the code comment here: https://github.com/flutter/flutter/blob/f71ceb77fe41e569c546ecd41d1b982ce2d0a0f2/packages/flutter/lib/src/painting/text_painter.dart#L702).

So when the range does not include a full cluster this defaults to the preferred line height which could be very different from the glyph height in some cases? But your test is passing. Let me look at how `GetRectsForRange` works.

Additionally if the text is empty I'm not sure what should `getRectForComposingRange(TextRange(start: 0, end: 1))` return since the range is OOB. I think we should do a proper boundary check instead of relying on undefined behavior.

Reply:
```cpp
      for (const GlyphPosition& gp : run.positions) {
        if (gp.code_units.start >= start && gp.code_units.end <= end) {
          left = std::min(left, static_cast<SkScalar>(gp.x_pos.start));
          right = std::max(right, static_cast<SkScalar>(gp.x_pos.end));
        } else if (gp.code_units.end == end) {
          // Calculate left and right when we are at
          // the last position of a combining character.
          glyph_length = (gp.code_units.end - gp.code_units.start) - 1;
          if (gp.code_units.start ==
              std::max<size_t>(0, (start - glyph_length))) {
            left = std::min(left, static_cast<SkScalar>(gp.x_pos.start));
            right = std::max(right, static_cast<SkScalar>(gp.x_pos.end));
          }
        }
      }
```

It looks like if the start handle is aligned with the family emoji it won't have the correct height?

Reply:
Oh I missed this behavior in my test. Currently the test selects 'def 👨‍👩‍👦 ' with the start handle at the 'd' and ending handle at the end of the grapheme cluster. 

![Screen Shot 2021-07-15 at 1 15 59 PM](https://user-images.githubusercontent.com/948037/125851863-b83271a3-4009-49a4-8b1e-8560a6d2741b.png)

The behavior you describe is what I see. At the beginning of the family emoji we don't have the correct height (preferredLineHeight is used in this case), and at the end we do have the correct height.

Reply:
I agree we should add a boundary check in `getRectForComposingRange` to make sure the range is valid for the current text.

Reply:
With the latest commit, I've gotten around this issue by getting the rectangles for the range of first/last full grapheme cluster of the given selection instead of only the first/last character of a given selection.

tight|max
---|---
![Screen Shot 2021-07-15 at 7 23 42 PM](https://user-images.githubusercontent.com/948037/125882375-ffd8bd6e-481f-4ccf-bd93-fee279e9c807.png)|![Screen Shot 2021-07-15 at 7 24 34 PM](https://user-images.githubusercontent.com/948037/125882386-5a7b64b4-f78c-4100-a726-41ede4ff3276.png)